src/partr.c

```diff
@@ -452,26 +469,41 @@ JL_DLLEXPORT jl_task_t *jl_task_get_next(jl_value_t *trypoptask, jl_value_t *q)
         jl_cpu_pause();
         jl_ptls_t ptls = ct->ptls;
         if (sleep_check_after_threshold(&start_cycles) || (!jl_atomic_load_relaxed(&_threadedregion) && ptls->tid == 0)) {
-            jl_atomic_store(&ptls->sleep_check_state, sleeping); // acquire sleep-check lock
-            if (!multiq_check_empty()) {
+            // acquire sleep-check lock
+            jl_atomic_store_relaxed(&ptls->sleep_check_state, sleeping);
+            jl_fence(); // add a sequenced-before edge between the sleep_check_state modify and get_next_task checks
```

Comment:
Yeah, this is annoying to describe properly. I probably want seq-cst happens-before here. But it is not with enqueuing of the task, since that might only be atomic release. And anyways, seq-cst stores only create a global ordering with other seq-cst operations and fences, but IIUC, does not necessarily enforce an ordering on other nearby operations.

Reply:
I *think* a somewhat concrete argument would be something like the following. Using the terminology by Lahav et al. (2017) (as standardese is hard...) we see that we have two _reads-before_ edges:

* *rb1*: `multiq_check_empty` -> `multiq_insert`
* *rb2*: `jl_atomic_load_relaxed(&ptls->sleep_check_state)` in `jl_wakeup_thread` -> `jl_atomic_store_relaxed(&ptls->sleep_check_state, sleeping)`

(Since they are inter-thread edges that start from a read and end with a write, they do not contribute to the usual happens-before edges.)

(Since *rb1* touches multiple memory locations, I guess it's more like a "bundle of edges".)

Let us also name the two fences sandwiching these edges:

* *F1*: [the first `jl_fence`](https://github.com/JuliaLang/julia/blob/c3f7396a35fd6c30754bccf0b293c3bdc56b808d/src/partr.c#L381) in `jl_wakeup_thread`
* *F2*: the `jl_fence` in `jl_task_get_next` that I quoted above

These fences do double duty of promoting the reads-before edges to the _partial SC_ relation edges:

* *sc1*: *F2* -> *rb1* -> *F1*
* *sc2*: *F1* -> *rb2* -> *F2*

Since both *sc1* and *sc2* are part of the partial SC relation that is acyclic under C+20 memory model, they can't form a cycle. We can use this to show that the following unwated execution is impossible:

* Initial state:
  * queue is empty
  * `sleep_check_state` is `not_sleeping`
* Thread 1 (dequeuer) does:
  * 1a: `jl_atomic_store_relaxed(&ptls->sleep_check_state, sleeping)`
  * 1b: `multiq_check_empty` returns true
* Thread 2 (enqueuer) does:
  * 2a: `multiq_insert`
  * 2b: `jl_atomic_load_relaxed(&ptls->sleep_check_state)` in `jl_wakeup_thread` returns `not_sleeping`

i.e., the dequeuer misses the enqueue and enqueuer misses the sleep state transition.

This is impossible because it forms a cycle

* 2b -> 1a (due to the value read in 2b)
* 1a -> 1b (sequenced-before)
* 1b -> 2a (due to the value read in 1b)
* 2a -> 2b (sequenced-before)

which is forbidden because both 1b -> 2a (*rb1*) and 2b -> 1a (*rb2*) are part of the partial SC relation, thanks to the SC fences.

But this argument is essentially a copy of the argument about the standard "store buffering litmus test" so I don't think it's helpful to expand the explanation this much.

Reply:
But I think it's *very* helpful to document which fences "pair up." Maybe we can use markdown footnote-like comment like this

```c
jl_fence(); // [^store_buffering_1]
```

in the first fence in `jl_wakeup_thread` and the fence of `jl_task_get_next` that I quoted above? We can then document which stores and loads are taken care of by this fence outside the function body (so that it's OK to write a somewhat lengthy comment).

```c
...
}
// [^store_buffering_1]: These fences are used to avoid the cycle 2b -> 1b -> 2a -> 2b where
// * Dequeuer:
//   * 1a: `jl_atomic_store_relaxed(&ptls->sleep_check_state, sleeping)`
//   * 1b: `multiq_check_empty` returns true
// * Enqueuer:
//   * 2a: `multiq_insert`
//   * 2b: `jl_atomic_load_relaxed(&ptls->sleep_check_state)` in `jl_wakeup_thread` returns `not_sleeping`
// i.e., the dequeuer misses the enqueue and enqueuer misses the sleep state transition.
```